### PR TITLE
Adds TestCoroutineDispatcher + LiveData coroutines support

### LINKED ISSUE #655: Test two LiveData consequent emissions in ShotViewModel
There's a test missing in ShortViewModelTest.

We want to test that the ViewModel emits two uiModel objects. The first one as a fast emission as soon as the VM is created and then, after a new coroutine in the background makes some processing, the second value.

Maybe there's a way we can control the Coroutine with a `ContinuationInterceptor`? We don't want to change the VM behaviour just for the sake of the tests.

### PR DESCRIPTION
**Pull request by manuelvicnt**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

Used the new experimental TestCoroutineDispatcher to fix #655.
And the new LiveData coroutines support to replace LiveData Transformations.

## :bulb: Motivation and Context

Before, we didn't have the option to stop / resume the CoroutinesDispatcher when injected with our CoroutinesDispatcherProvider so we couldn't test logic that relies on that.

## :green_heart: How did you test it?
Unit tests

## :pencil: Checklist

- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing

## :crystal_ball: Next steps
When the API hits stable, we'll migrate all the tests to use this. runBlocking in tests should migrate to runBlockingTest.

### CODE REVIEW

**Comment by florina-muntenescu** [comment]
The fail seems genuine:
```
 java.lang.RuntimeException: Duplicate class kotlinx.coroutines.test.TimedRunnable found in modules kotlinx-coroutines-core-1.2.2.jar (org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.2) and kotlinx-coroutines-test-1.2.2.jar (org.jetbrains.kotlinx:kotlinx-coroutines-test:1.2.2)
``` 
Could you fix it, rebase and if everything passes, merge?

File: app/src/main/java/io/plaidapp/ui/HomeViewModel.kt [hunk 1]
```diff
@@ -109,10 +110,10 @@ class HomeViewModel(
 
     fun getFeed(columns: Int): LiveData<FeedUiModel> {
         return Transformations.switchMap(feedData) {
-            // TODO move this on a background thread
-            //  https://github.com/nickbutcher/plaid/issues/658
-            expandPopularItems(it, columns)
-            return@switchMap MutableLiveData(FeedUiModel(it))
+            return@switchMap liveData(viewModelScope.coroutineContext + dispatcherProvider.computation) {
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeViewModel.kt]: Can we use the `LiveData.switchMap` ext fun here?

File: build.gradle [hunk 2]
```diff
@@ -43,7 +42,7 @@ buildscript { scriptHandler ->
             'kotlin'             : '1.3.41',
```
Comment (nickbutcher) [inline build.gradle]: `1.3.50` While we're here?

File: dribbble/src/test/java/io/plaidapp/dribbble/ui/shot/ShotViewModelTest.kt [hunk 3]
```diff
@@ -143,6 +155,26 @@ class ShotViewModelTest {
         assertEquals(id, shotId)
     }
 
+    @Test
+    fun loadShot_emitsTwoUiModels() = testCoroutineDispatcher.runBlockingTest {
```
Comment (nickbutcher) [inline dribbble/src/test/java/io/plaidapp/dribbble/ui/shot/ShotViewModelTest.kt]: Nice!

File: search/src/main/java/io/plaidapp/search/ui/SearchViewModel.kt [hunk 4]
```diff
@@ -43,10 +45,12 @@ class SearchViewModel(
 
     private val searchQuery = MutableLiveData<String>()
 
-    private val results = Transformations.switchMap(searchQuery) {
-        loadSearchData = LoadSearchDataUseCase(factories, it)
-        loadMore()
-        return@switchMap loadSearchData?.searchResult
+    private val results: LiveData<List<PlaidItem>> = Transformations.switchMap(searchQuery) {
```
Comment (nickbutcher) [inline search/src/main/java/io/plaidapp/search/ui/SearchViewModel.kt]: Ditto ext fun.

File: test_shared/src/main/java/io/plaidapp/test/shared/MainCoroutineRule.kt [hunk 5]
```diff
@@ -18,26 +18,31 @@ package io.plaidapp.test.shared
...
-        singleThreadExecutor.shutdownNow()
         Dispatchers.resetMain()
+        testDispatcher.cleanupTestCoroutines()
     }
 }
+
+@ExperimentalCoroutinesApi
+fun MainCoroutineRule.runBlocking(block: suspend () -> Unit) = this.testDispatcher.runBlockingTest {
+    block()
+}
```
Comment (nickbutcher) [inline test_shared/src/main/java/io/plaidapp/test/shared/MainCoroutineRule.kt]: nit: newline

File: test_shared/src/main/java/io/plaidapp/test/shared/MainCoroutineRule.kt [hunk 6]
```diff
@@ -18,26 +18,31 @@ package io.plaidapp.test.shared
...
 import kotlinx.coroutines.test.setMain
 import org.junit.rules.TestWatcher
 import org.junit.runner.Description
-import java.util.concurrent.Executors
 
 @ExperimentalCoroutinesApi
-class CoroutinesMainDispatcherRule : TestWatcher() {
-
-    private val singleThreadExecutor = Executors.newSingleThreadExecutor()
+class MainCoroutineRule(
```
Comment (nickbutcher) [inline test_shared/src/main/java/io/plaidapp/test/shared/MainCoroutineRule.kt]: Class comment

**Pull request merged**